File: core/features/rating.py
"""Module for working with note ratings."""

import logging
from pathlib import Path
from typing import Optional

logger = logging.getLogger(__name__)


def update_rating(filepath: Path, rating: int) -> bool:
    """
    Update the rating field in a note's frontmatter.

    Args:
        filepath: Path to the note file
        rating: New rating value (1-5)

    Returns:
        True if successful, False otherwise
    """
    try:
        if not filepath.exists():
            logger.error(f"File not found: {filepath}")
            return False

        # Read file content
        content = filepath.read_text(encoding="utf-8")

        new_content = update_rating_impl(content, rating)

        # Write back to file
        if new_content:
            filepath.write_text(new_content, encoding="utf-8")
            logger.info(f"Successfully updated rating to {rating} in {filepath}")
        else:
            logger.warning(f"Error while updating rating to {rating} in {filepath}")
        return True

    except Exception as e:
        logger.error(f"Error updating rating in {filepath}: {e}")
        return False


def update_rating_impl(content: str, rating: int) -> Optional[str]:
    # Split by frontmatter delimiters
    parts = content.split("---")
    if len(parts) < 3:
        logger.error("Invalid frontmatter format in note")
        return None

    # Parse frontmatter (between first and second ---)
    frontmatter = parts[1]
    lines = frontmatter.split("\n")

    # Update or add rating field
    rating_found = False
    updated_lines: list[str] = []
    for line in lines:
        if line.strip().startswith("Оценка:"):
            updated_lines.append(f"Оценка: {rating}")
            rating_found = True
        else:
            updated_lines.append(line)

    # If rating field not found, add it
    if not rating_found:
        # Insert before the last empty line if exists
        if updated_lines and updated_lines[-1] == "":
            updated_lines.insert(-1, f"Оценка: {rating}")
        else:
            updated_lines.append(f"Оценка: {rating}")

    # Reconstruct content
    parts[1] = "\n".join(updated_lines)
    new_content = "---".join(parts)

    # Write back to file
    return new_content


def get_rating(filepath: Path) -> Optional[int]:
    """
    Extract the rating value from a note's frontmatter.

    Args:
        filepath: Path to the note file

    Returns:
        Rating value as integer, or None if not found or on error
    """
    try:
        if not filepath.exists():
            logger.error(f"File not found: {filepath}")
            return None

        # Read file content
        content = filepath.read_text(encoding="utf-8")

        return get_rating_impl(content)

    except Exception as e:
        logger.error(f"Error reading rating from {filepath}: {e}")
        return None


def get_rating_impl(content: str) -> Optional[int]:
    # Split by frontmatter delimiters
    parts = content.split("---")
    if len(parts) < 3:
        logger.warning("Invalid frontmatter format in note")
        return None

    # Parse frontmatter (between first and second ---)
    frontmatter = parts[1]
    lines = frontmatter.split("\n")

    # Find rating field
    for line in lines:
        if line.strip().startswith("Оценка:"):
            # Extract value after colon
            rating_str = line.split(":", 1)[1].strip()
            try:
                rating = int(rating_str)
                return rating
            except ValueError:
                logger.warning(f"Invalid rating value in note: {rating_str}")
                return None

    # Rating field not found
    return None

File: core/features/rating.go
package features

import (
	"context"
	"notes-bot/internal/telemetry"
	"strconv"
	"strings"

	"go.uber.org/zap"
)

func UpdateRatingImpl(ctx context.Context, content string, rating int) (string, bool) {
	ctx, span := telemetry.StartSpan(ctx)
	defer span.End()

	parts := strings.Split(content, "---")
	if len(parts) < 3 {
		logger.Error("invalid frontmatter format in note")
		return "", false
	}

	// Handle both 2-delimiter and 3-delimiter (Obsidian) formats
	// parts[0] = "" (before first ---)
	// parts[1] = frontmatter
	// parts[2] = content (may contain trailing --- in Obsidian format)
	frontmatter := parts[1]
	var contentSuffix string

	// If there's a third delimiter, frontmatter might contain part of content
	if len(parts) >= 4 {
		// In Obsidian format: ---frontmatter---content---
		// parts[1] = "frontmatter\n" and parts[2] = "content\n---"
		// We need to extract just the frontmatter part and preserve the rest
		frontmatterLines := strings.Split(frontmatter, "\n")
		var cleanFrontmatter []string
		var contentStart []string
		inFrontmatter := true

		for _, line := range frontmatterLines {
			// Stop if we hit what looks like content (starts with - [ ] or - [x])
			if inFrontmatter && strings.HasPrefix(strings.TrimSpace(line), "- [") {
				inFrontmatter = false
			}
			if inFrontmatter {
				cleanFrontmatter = append(cleanFrontmatter, line)
			} else {
				contentStart = append(contentStart, line)
			}
		}
		frontmatter = strings.Join(cleanFrontmatter, "\n")
		// Reconstruct the content part that was in frontmatter
		if len(contentStart) > 0 {
			contentSuffix = strings.Join(contentStart, "\n")
			if !strings.HasPrefix(contentSuffix, "\n") {
				contentSuffix = "\n" + contentSuffix
			}
		}
	}

	lines := strings.Split(frontmatter, "\n")

	ratingFound := false
	updatedLines := make([]string, 0, len(lines))

	for _, line := range lines {
		if strings.HasPrefix(strings.TrimSpace(line), "Оценка:") {
			ratingFound = true
			updatedLines = append(updatedLines, "Оценка: "+strconv.Itoa(rating))
		} else {
			updatedLines = append(updatedLines, line)
		}
	}

	if !ratingFound {
		if len(updatedLines) > 0 && updatedLines[len(updatedLines)-1] == "" {
			updatedLines = append(updatedLines[:len(updatedLines)-1], "Оценка: "+strconv.Itoa(rating), "")
		} else {
			updatedLines = append(updatedLines, "Оценка: "+strconv.Itoa(rating))
		}
	}

	parts[1] = strings.Join(updatedLines, "\n") + contentSuffix
	return strings.Join(parts, "---"), true
}

func GetRatingImpl(ctx context.Context, content string) *int {
	ctx, span := telemetry.StartSpan(ctx)
	defer span.End()

	parts := strings.Split(content, "---")
	if len(parts) < 3 {
		logger.Error("invalid frontmatter format in note")
		return nil
	}

	// Handle both 2-delimiter and 3-delimiter (Obsidian) formats
	// parts[0] = "" (before first ---)
	// parts[1] = frontmatter
	// parts[2] = content (may contain trailing --- in Obsidian format)
	frontmatter := parts[1]
	// If there's a third delimiter, frontmatter might contain part of content
	// Find the actual end of frontmatter by looking for the second ---
	if len(parts) >= 4 {
		// In Obsidian format: ---frontmatter---content---
		// parts[1] = "frontmatter\n" and parts[2] = "content\n---"
		// We need to extract just the frontmatter part
		frontmatterLines := strings.Split(frontmatter, "\n")
		var cleanLines []string
		for _, line := range frontmatterLines {
			// Stop if we hit what looks like content (starts with - [ ] or - [x])
			if strings.HasPrefix(strings.TrimSpace(line), "- [") {
				break
			}
			cleanLines = append(cleanLines, line)
		}
		frontmatter = strings.Join(cleanLines, "\n")
	}

	lines := strings.SplitSeq(frontmatter, "\n")

	for line := range lines {
		if !strings.HasPrefix(strings.TrimSpace(line), "Оценка:") {
			continue
		}
		ratingStr := strings.TrimSpace(strings.SplitN(line, ":", 2)[1])
		// Empty rating is normal - note just doesn't have a rating yet
		if ratingStr == "" {
			return nil
		}
		rating, err := strconv.Atoi(ratingStr)
		if err != nil {
			logger.Warn("invalid rating value", zap.String("value", ratingStr))
			return nil
		}
		return &rating
	}

	return nil
}
